### ISSUE TITLE
Option for wait_all to propagate exceptions eagarly

### ISSUE THREAD

**Issue by Timmmm** [jbaldwin/libcoro#280]
Hi, we're using libcoro for test stimulus generation in an SV simulation. It's working perfectly (thanks!), *except* for exceptions (heh). Ideally I would like any exception thrown in the coroutine to immediately get propagated into the non-coroutine that resumed it. But libcoro doesn't seem to have an option for that - it stores the exception until the coroutine is completely finished and then rethrows it when you get the promise result.

That's why I added this code to our event loop to check if the stimulus is finished and rethrow its exception if there is one.

```
    // Check if the stimulus is finished.
    if (m_task.is_ready()) {
        // Rethrow exception if there was one.
        m_task.promise().result();
    }
```

It works for simple cases. Unfortunately it breaks down with `when_all`. We might have infinite stimulus, e.g. like this:

```
task<void> stimulus(Agent& agent) {
  co_await when_all(random_writes(agent), random_reads(agent));
  co_return;
}

task<void> random_writes(Agent& agent) {
  for (;;) {
    co_await agent.write(...);
  }
}

task<void> random_reads(Agent& agent) {
  for (;;) {
    co_await agent.read(...);
  }
}
```

This stimulus also works great, except when `write()` throws an exception, it doesn't get propagated through `when_all()`, and since `when_all()` never actually returns we never see the exception.

Like `task` it just stores the exception until it is finished:

    auto unhandled_exception() noexcept -> void { m_exception_ptr = std::current_exception(); }

I think `when_all` style functions in other languages have an option to say whether to store exceptions or to rethrow them as soon as they occur, e.g. [Python](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.wait).

I dunno if `when_all`'s implementation is correct even if the intention is to store them, since it can only rethrow one exception. If it captures multiple exceptions they'll just overwrite each other - possibly in a random order.

**Comment by jbaldwin** [comment]
Lets add a test for this, because it definitely is missing when I look in `test_when_all.cpp`. I think the intention is that each `coro::task` should be catching and storing its own exception and the exception you pointed out would be something goes wrong in `when_all` itself.. but its been a while since I've been in this code so lets try it out.

The behavior _should_ be that when you are iterating the `task` results of `when_all` each individual `task` can throw its own exception.

**Comment by jbaldwin** [comment]
Since `coro::when_all` is supposed to complete all the tasks its given so it isn't possible to just stop all the executing tasks and return with the exception.

I think its better to think of each task having its _own_ possible exception and they are unrelated. I posted a test that shows how this works in #281 if you want to take a look at the expected behavior.

I think a `coro::when_any` would probably behavior more like what you want, except that my original design would be that the first task to produce any result would then cancel the rest of the tasks. Result being a the return type T or an exception.

**Comment by Timmmm** [comment]
I would expect `when_all` to (optionally) have different behaviour when one of the tasks throws an exception though.

To put it another way, what code would you expect a user to write if they want to run multiple tasks to completion, but get immediately notified if any throws an exception? That is possible (and often the default) in every other async system I know of.

**Comment by jbaldwin** [comment]
Hmm, that doesn't seem right to me. Looking at say C#'s [Task.WhenAll](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.whenall?view=net-8.0) it waits for all the tasks its given before its own task is completed, which is exactly how `coro::when_all` works. The behavior you're asking for isn't how I'm familiar with `when_all` type functionality works.

Do you have an example of a `WhenAll` that does behave the way you are expecting it to that I could look at?

There are probably a lot of ways to do what you want, but libcoro doesn't have any pre-built structure to do so. Wrapping your tasks in another task that catches the throw and does what you want once it throws exactly once via maybe a `coro::event` or something else might work? I could probably make an example later today.

**Comment by Timmmm** [comment]
Ha it seems like we are familiar with different languages and they behave differently. JavaScript's `Promise.all` rethrows the first exception immediately and Python's equivalent is configurable - either rethrows the first exception or collect them all. Rust's [`try_join`](https://docs.rs/futures-util/latest/futures_util/macro.try_join.html) will return as soon as any task returns an error.

C# seems to be the exception here. I think you would need to have a loop around `WhenAny`, which is fine, but that doesn't work in libcoro because it doesn't have `when_any`.

> Wrapping your tasks in another task that catches the throw and does what you want once it throws exactly once via maybe a coro::event or something else might work?

I'm not sure if that will work... (haven't learnt `coro::event` yet!) but it sounds like it would make the API worse. Coroutines are pretty hardcore C++ already; I don't think I want to add any extra footguns for my colleagues. I think I would rather just patch our copy of libcoro tbh.

**Comment by jbaldwin** [comment]
Yeah it looks like Promise.all() calls reject on the first failure but Promise.allSettled() works like coro::when_all right now.

Rust has join! and try_join! I'm familiar with both. coro::when_all is like join!.

You're right this is definitely common so I think there's two useful constructs to add to libcoro:

1) when_all with stop execution of all tasks on an exception. We can call this coro::when_all_return_on_exception until a better name comes to mind, or add a parameter for how when _all should behave. This is pretty easy to add in a single threaded app, if the current task executing throws just return and never start the remaining tasks. However if you are using coro::when_all with a coro::thread_pool then the tasks can all be executing in parallel on different threads, if one of them throws the context can catch it in the order it started executing (it doesn't constantly loop over them it just waits in the given order) and then return. Using the coro::io_scheduler it might be possible to work around this in a parallel task execution scenario but I'd need to probably play around with it and would most likely require a coro::io_scheduler to be passed in as a parameter to guarantee it runs well.

2) when_any just return the first result to complete. Again super easy in single threaded that is looping over the tasks and awaiting them, but multi threaded could be a big issue for all the same reasons as pointed out in the prior idea.

I still think to do these features properly for a multi threaded context a std::stop_token like object should be provided to basically every task, or at least optionally. Then the context can stop any tasks that are running in parallel on a thread pool, or at least tell the user they should stop if they are checking it. I would probably do this by wrapping each user task in a private class that when an exception is thrown it gets triggered to run once, marks the stop token and then returns the result on the coro::when_all_return_on_exception or coro:: when_any.

**Comment by jbaldwin** [comment]
I'm finally a bit free to try and implement some code for this, but it looks like some emscripten stuff is failing on the build right now :(

**Comment by Timmmm** [comment]
I think this got closed accidentally?

**Comment by jbaldwin** [comment]
Oh yeah, the test PR was linked so GitHub closed it when I merged it in, I'll reopen.

**Comment by jbaldwin** [comment]
Ok so there is a major lifetime issue with `when_any` with the current libcoro design of taking ownership of the tasks when executing. The function prototype (roughly):

```c++
coro::when_any(range_type&& awaitables) -> task<return_type>
```

Taking ownership means the other N -1 tasks 'die' after `when_any` returns and ends up segfaulting if they continue executing.

Some options could be:

1. Wait for all tasks to finish like `coro::when_all` but only return the first result. This seems like it kind of defeats the purpose of the function but would be safe, it would require the user to actually cancel tasks to get the desired behavior which seems un-ideal.
2. Take the tasks by reference and force the user to maintain proper lifetimes, I dislike this since it has pitfalls but would in theory work if the user uses it correctly. There is still a wrapper promise which waits for 1/all tasks and would need to self destruct probably via ref counting upon all tasks completing or cancelling.
3. Require a `coro::task_container` parameter that the tasks would have their lifetime managed by. This would be safer than (2) and _probably_ easier to get right but introduces a fairly heavy runtime/lifetime management object.

**Comment by Timmmm** [comment]
I agree with your points. I'm not sure I understand the lifetime issues involved, but have you considered simply returning all of the tasks back to the caller, together with an index saying which one finished? Though I guess that would still break if you discard the returned tasks?

If you take them by reference is the lifetime issue just that you have to keep the `task<>` objects alive until `when_any` returns? If so I think that's probably acceptable. Obviously it sucks compared to Rust, but you kind of have to live with it in C++. Perhaps it's acceptable if well documented, and maybe even encoded in the name (`when_any_by_ref()` or whatever)?

**Comment by jbaldwin** [comment]
I think returning with an index or taking by ref have the same issue, the user needs to keep the tasks alive until they stop executing, not just for as long as `when_any` returns since the unfinished tasks could still be running.

**Comment by Timmmm** [comment]
Ok I confess I'm not using libcoro with any of its executors so I don't really understand the limitations there - I'm manually `.resume()`ing a single top-level `task<>` in a single thread so it doesn't matter if tasks are dropped (as far as I know anyway; how I wish C++ had a borrow checker).

I guess I'll stick with our modified version for now.

**Comment by jbaldwin** [comment]
Yeah sorry I'm not really coming up with a good generic way to integrate it to the library yet that is a good all around fit, but I'd really like to.

Another thought I had was maybe force a user to make an executor just like tokio does all the time in rust. That would include use cases like yours and then all tasks could probably be managed correctly. I don't really like forcing this on the user since it's really nice right now to just jump in and start using tasks via sync wait. It would also be a huge change in how tasks get scheduled for all use cases.

**Comment by jbaldwin** [comment]
@Timmmm I've had a new thought on how to do this, I've recently introduced a special new task `coro::detail::task_self_deleting` which doesn't suspend the coroutine upon completion and relies on the spec to destroy the coroutine frame.

I'm wondering if a solution like:

```c+++
template<typename return_type>
auto coro::when_any(vector<task<return_type>> tasks, coro::event cancellation_token) -> coro::task<return_type>
{
    event e{};

    vector<task_self_deleting> orphaned;
    for(auto t : tasks)
        orphaned.emplace_back(make_task_self_deleting(std::move(t), e);

    // where the first task that is orphaned sets an event to trigger this coroutine to wake-up.
    co_await e;

    // send cancellation to other tasks? they must check for it though, and they must have had to pass it into their
    // tasks to be able to even have access to it.
    cancellation_token.set();

   /// the first task also sets the return value somewhere special that can be picked up, possibly on a modified event object.
    co_return e.value;
}

```

This will just return the first result but doesn't send any kind of cancellation to the other tasks -- but perhaps this is a good first step. I _think_ this solves the lifetime issues that we ran into before using just `coro::task` but I need to give this a shot. I figured I'd write down my thoughts before trying it though.

**Comment by jbaldwin** [comment]
@Timmmm I've got a working version of `coro::when_any` working on the original ticket you opened #279 . I'm going to close this one out.
